src/usr/pnor/sfc_memboot_ast2400.C

```diff
@@ -0,0 +1,180 @@
+#include "norflash.H"
+#include <util/align.H>
+
+/*****************************************************************************/
+// C o n s t a n t s
+/*****************************************************************************/
+
+
+
+
+/*****************************************************************************/
+// G l o b a l s
+/*****************************************************************************/
+
+// Initialized in pnorrp.C
+extern trace_desc_t* g_trac_pnor;
+
+/*****************************************************************************/
+// M e t h o d s
+/*****************************************************************************/
+
+
+namespace PNOR {
+
+/**
+ * @brief Wrapper for device driver constructor
+ */
+errlHndl_t create_SfcDD( SfcDD*& o_sfc,
+                         TARGETING::Target* i_proc )
+{
+    errlHndl_t l_err = NULL;
+    TRACFCOMP( g_trac_pnor, "Creating SfcMembootAST2400 object" );
+    o_sfc = new SfcMembootAST2400( l_err, i_proc );
+    return l_err;
+}
+
+};
+
+/**
+ * @brief Constructor
+ */
+SfcMembootAST2400::SfcMembootAST2400( errlHndl_t& o_err,
+                        TARGETING::Target* i_proc )
+: SfcDD(o_err,i_proc)
+{
+}
+
+
+/**
+ * @brief Read data from the flash
+ */
+errlHndl_t SfcMembootAST2400::readFlash( uint32_t i_addr,
+                                  size_t i_size,
+                                  void* o_data )
+{
+    errlHndl_t l_err = NULL;
+    TRACDCOMP( g_trac_pnor, ENTER_MRK"SfcMembootAST2400::readFlash> i_addr=0x%.8x, i_size=0x%.8x",  i_addr, i_size );
+
+    do{
+        uint32_t* word_ptr = static_cast<uint32_t*>(o_data);
+        uint32_t word_size = (ALIGN_4(i_size))/4;
+        for( uint32_t words_read = 0;
+             words_read < word_size;
+             words_read ++ )
+        {
+            //Read directly from MMIO space
+            uint32_t lpc_addr = LPC_SFC_MMIO_OFFSET | (i_addr + words_read*4);
+            size_t reg_size = sizeof(uint32_t);
+
+            l_err = deviceOp( DeviceFW::READ,
+                              iv_proc,
+                              &(word_ptr[words_read]),
+                              reg_size,
+                              DEVICE_LPC_ADDRESS(LPC::TRANS_FW,
+                                                 lpc_addr) );
+            if( l_err ) {  break; }
+        }
+        if( l_err ) {  break; }
+    }while(0);
+
+    TRACDCOMP( g_trac_pnor, EXIT_MRK"SfcMembootAST2400::readFlash> err=%.8X", ERRL_GETEID_SAFE(l_err) );
+    return l_err;
+}
+
+
+/**
+ * @brief Write data into flash
+ */
+errlHndl_t SfcMembootAST2400::writeFlash( uint32_t i_addr,
+                                   size_t i_size,
+                                   void* i_data )
+{
+    errlHndl_t l_err = NULL;
+    TRACDCOMP( g_trac_pnor, ENTER_MRK"SfcMembootAST2400::writeFlash> i_addr=0x%.8x, i_size=0x%.8x", i_addr, i_size );
+    TRACFCOMP( g_trac_pnor, "SfcMembootAST2400::writeFlash> Not supported for memboot");
```

Comment:
On Tue, 24 Feb 2015 14:16:31 Patrick Williams wrote:
> > +
> > +    TRACDCOMP( g_trac_pnor, EXIT_MRK"SfcMembootAST2400::readFlash>
> > err=%.8X", ERRL_GETEID_SAFE(l_err) ); +    return l_err;
> > +}
> > +
> > +
> > +/**
> > + * @brief Write data into flash
> > + */
> > +errlHndl_t SfcMembootAST2400::writeFlash( uint32_t i_addr,
> > +                                   size_t i_size,
> > +                                   void* i_data )
> > +{
> > +    errlHndl_t l_err = NULL;
> > +    TRACDCOMP( g_trac_pnor, ENTER_MRK"SfcMembootAST2400::writeFlash>
> > i_addr=0x%.8x, i_size=0x%.8x", i_addr, i_size ); +    TRACFCOMP(
> > g_trac_pnor, "SfcMembootAST2400::writeFlash> Not supported for memboot");
> Want to make this wrapped by a static-bool to make it one time only?

Good catch. I have an updated patch that implements the write but I couldn't 
find where Hostboot writes to flash so I wasn't able to test it. Does Hostboot 
write to flash at the moment or is this something it might do in future? What 
would be the best way of testing flash writes? Or should I just submit it with 
the write implementation and hope that it works/leave the testing to someone 
else?

Thanks.

- Alistair

> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/open-power/hostboot/pull/9/files#r25298839